components/storage/blockdevice/COMPONENT_OSPIF/OSPIFBlockDevice.cpp

```diff
@@ -0,0 +1,1908 @@
+#define DTROSPIF_INST_READ_OCTA_DTR  0xEE // Read data in DOPI mode
+#define OSPIF_INST_WRCR2  0x72 // Write Configuration Register2
+#define OSPIF_INST_RDCR2  0x71 // Read Configuration Register2
+
+//Macronix flash 4-Byte address instructions
+#define OSPIF_INST_READ_4B  0x13 // Read data with 4 byte address
+#define OSPIF_INST_PROG_4B  0x12 // Page program with 4 byte address
+
+// Device-specific instructions
+#define OSPIF_INST_ULBPR 0x98 // Clear all write-protection bits in the Block-Protection register
+#define OSPIF_INST_RDCR  0x15 // Read the two control registers
+
+// Default read/legacy erase instructions
+#define OSPIF_INST_READ_DEFAULT          0x03
+#define OSPIF_INST_LEGACY_ERASE_DEFAULT  OSPI_NO_INST
+
+// Default status register 2 read/write instructions
+#define OSPIF_INST_WSR2_DEFAULT    OSPI_NO_INST
+#define OSPIF_INST_RSR2_DEFAULT    0x35
+
+// Default 4-byte extended addressing register write instruction
+#define OSPIF_INST_4BYTE_REG_WRITE_DEFAULT OSPI_NO_INST
+
+
+// Length of data returned from RDID instruction
+#define OSPI_RDID_DATA_LENGTH 3
+
+
+/* Init function to initialize Different Devices CS static list */
+static PinName *generate_initialized_active_ospif_csel_arr();
+// Static Members for different devices csel
+// _devices_mutex is used to lock csel list - only one OSPIFBlockDevice instance per csel is allowed
+SingletonPtr<PlatformMutex> OSPIFBlockDevice::_devices_mutex;
+int OSPIFBlockDevice::_number_of_active_ospif_flash_csel = 0;
+PinName *OSPIFBlockDevice::_active_ospif_flash_csel_arr = generate_initialized_active_ospif_csel_arr();
+
+/********* Public API Functions *********/
+/****************************************/
+OSPIFBlockDevice::OSPIFBlockDevice(PinName io0, PinName io1, PinName io2, PinName io3, PinName io4, PinName io5, PinName io6, PinName io7, PinName sclk, PinName csel, PinName dqs,
+                                   int clock_mode, int freq)
+    : _ospi(io0, io1, io2, io3, io4, io5, io6, io7, sclk, csel, dqs, clock_mode), _csel(csel), _freq(freq),
+      _init_ref_count(0),
+      _is_initialized(false)
+{
+    _unique_device_status = add_new_csel_instance(csel);
+
+    if (_unique_device_status == 0) {
+        tr_debug("Adding a new OSPIFBlockDevice csel: %d", (int)csel);
+    } else if (_unique_device_status == -1) {
+        tr_error("OSPIFBlockDevice with the same csel(%d) already exists", (int)csel);
+    } else {
+        tr_error("Too many different OSPIFBlockDevice devices - max allowed: %d", OSPIF_MAX_ACTIVE_FLASH_DEVICES);
+    }
+
+    // Initialize parameters
+    _sfdp_info.bptbl.legacy_erase_instruction = OSPIF_INST_LEGACY_ERASE_DEFAULT;
+    _sfdp_info.bptbl.device_size_bytes = 0;
+    _sfdp_info.smptbl.regions_min_common_erase_size = 0;
+    _sfdp_info.smptbl.region_cnt = 1;
+    _sfdp_info.smptbl.region_erase_types_bitfld[0] = SFDP_ERASE_BITMASK_NONE;
+
+    // Until proven otherwise, assume no quad enable
```

Comment:
"// Until proven otherwise, assume no octal enable"?

Reply:
NO, we are planning to combine implementations between QSPI and OSPI.

Reply:
That's great :)